tools/serverpod_cli/lib/src/language_server/language_server.dart

```diff
@@ -39,36 +111,56 @@ Future<void> runLanguageServer() async {
       );
     });
 
-    var diagnostics = _validateYamlProtocol(
-      contentChanges.last.text,
-      params.textDocument.uri.toString(),
+    statefulAnalyzer.validateProtocol(
+      contentChanges.first.text,
+      params.textDocument.uri,
     );
 
-    connection.sendDiagnostics(
-      PublishDiagnosticsParams(
-        diagnostics: diagnostics,
-        uri: params.textDocument.uri,
-      ),
-    );
+    // This can be optimised to only validate the files we know have related errors.
+    statefulAnalyzer.validateAll();
   });
 
   await connection.listen();
 }
 
-List<Diagnostic> _validateYamlProtocol(String yaml, String sourcePath) {
-  var collector = CodeGenerationCollector();
+Future<GeneratorConfig> _initializeAnalyzer(
+  StatefulAnalyzer statefulAnalyzer,
+  Uri rootUri,
+) async {
+  var rootDir = Directory.fromUri(rootUri);
+
+  var serverDir = findServerDirectory(rootDir);
+  if (serverDir == null) {
+    throw Exception('No Serverpod server directory found.');
+  }
 
-  var analyzer = SerializableEntityAnalyzer(
-    yaml: yaml,
-    sourceFileName: sourcePath,
-    outFileName: '', // Only needed for real codegen
-    subDirectoryParts: [], // Only needed for real codegen
-    collector: collector,
-  );
+  var config = await GeneratorConfig.load(serverDir.path);
+  var yamlSources =
+      await ProtocolHelper.loadProjectYamlProtocolsFromDisk(config!);
 
-  analyzer.analyze();
+  statefulAnalyzer.initialValidation(yamlSources);
 
-  var diagnostics = collector.errors.where((e) => e.span != null).map((error) {
+  return config;
+}
+
+Directory? findServerDirectory(Directory rootDir) {
+  if (isServerDirectory(rootDir)) return rootDir;
+
+  var childDirs = rootDir.listSync().where(
+        (dir) => isServerDirectory(Directory.fromUri(dir.uri)),
+      );
+
+  if (childDirs.isNotEmpty) {
+    return Directory.fromUri(childDirs.first.uri);
```

Comment:
Question: This will only return the first root directory right? Are we OK with that?

Reply:
I made the assumption that you only have one Server project in the folder you opened. If we want to support several server projects we should implement the workspace feature.

Reply:
Seems ok, but I saw that you added an issue for this. Probably want to fix in the near future.